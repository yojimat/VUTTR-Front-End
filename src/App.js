import React, { Component } from "react";
import Vuttr from "./componentes/Vuttr";
import ModalVuttr from "./componentes/ModalVuttr";
import AddTool from "./componentes/AddTool";
import RemoveTool from "./componentes/RemoveTool";

class App extends Component {
	constructor() {
  		super();
  		this.ref = React.createRef();
  		this.state = {
    		modalAdicionarAberta: false,
    		modalDeleteAberta: false,
    		toolsList: [],
    		toolIdDelete: null,
    		nomeTool: "",
    		tagFiltro: false,
    		isLoading: false,
    		respostaFetchUsuario: ""
  		};
	}

	componentDidMount() {
		this.getListaTools();
	}

	async getListaTools() {
		const dados =  await fetch("/tools", {
			method: "get",
			headers: {
				'Content-Type': 'application/json'
			}
		})
		.then(resp => resp.json())
		.then(data => {
			return data;
		})
		.catch(error => console.log(error));

		if (dados !== undefined) {
			this.setState({ toolsList: dados });
		}	
	}

	async getListaToolsBuscaGlobal(busca) {
		const dados =  await fetch(`/tools?q=${busca}`, {
			method: "get",
			headers: {
				'Content-Type': 'application/json'
			}
		})
		.then(resp => resp.json())
		.then(data => {
			return data;
		})
		.catch(error => console.log(error));

		if (dados !== undefined) {
			this.setState({ toolsList: dados });
		}	
	}

	async getListaToolsBuscaTag(tag) {
		const dados =  await fetch(`/tools?tags_like=${tag}`, {
			method: "get",
			headers: {
				'Content-Type': 'application/json'
			}
		})
		.then(resp => resp.json())
		.then(data => {
			return data;
		})
		.catch(error => console.log(error));

		if (dados !== undefined) {
			this.setState({ toolsList: dados });
		}	
	}

	//ArrowFunction como método para receber valorPesquisa, evitando erro de propagação.
	pesquisaFiltradasTools = (valorPesquisa) => {
		const { tagFiltro } = this.state;
		if (tagFiltro) {
			this.getListaToolsBuscaTag(valorPesquisa);
		} else {
			this.getListaToolsBuscaGlobal(valorPesquisa);
		}
	}

	toggleTagFiltro = (refValue) => {
		this.setState({ tagFiltro: !this.state.tagFiltro }, () =>{
			this.pesquisaFiltradasTools(refValue);
		});
	}

	toggleModal () {
  		this.setState({ modalAdicionarAberta: !this.state.modalAdicionarAberta });
	}

	//ArrowFunction como método para receber id, evitando erro de propagação.
	toggleModalDelete = (item) => {
  		this.setState(
  			{ modalDeleteAberta: !this.state.modalDeleteAberta },
  			() => {
  				if (this.state.modalDeleteAberta === true && item !== undefined) { 
  					this.setState({
  						toolIdDelete: item.id,
  						nomeTool: item.title
  					}); 
  				} else { 
					this.setState({
						toolIdDelete: null,
						nomeTool: ""
					});
					this.loaderStatus("mensagemRecebida");
				}
  			}
  		);
	}
	//Retornar resposta para faciliar testes.
	deleteTool() {
		const { toolIdDelete } = this.state;

		this.loaderStatus();

		fetch(`/tools/${toolIdDelete}`,{
			method: "delete",
			headers: {
				'Content-Type': 'application/json'
			}
		})
		.then(resp => {
			this.getListaTools();

			if (resp.status === 200) { this.loaderStatus("Sucesso!") }
				else { this.loaderStatus(".") }
		})
		.catch(error => console.log(error));
	}

	adicionarTool = (form) => {
		return fetch("/tools/",{
			method: "post",
			headers: {
				'Content-Type': 'application/json'
			},
			body: JSON.stringify(form)      
		})
		.then(resp => {
			this.getListaTools();

			return resp.json();
		})
		.catch(error => console.log(error));
	}

	//Adicionar teste.
	loaderStatus = (mensagem = "loading") => {
		if (mensagem === "loading") { this.setState({ isLoading: true }) } 
			else if ( mensagem === "mensagemRecebida") { 
				this.setState({ 
					isLoading: false,
					respostaFetchUsuario: ""
				}) 
			} else { this.setState({ respostaFetchUsuario: mensagem }) }
	}

	render() {
		const { modalAdicionarAberta, modalDeleteAberta, toolsList, nomeTool, isLoading, respostaFetchUsuario } = this.state;

		return(
			<div>
				<Vuttr 
					abrirModal={() => this.toggleModal()} 
					abrirModalDelete={ this.toggleModalDelete }
					toolsList={ toolsList }
					ref={this.ref}
					pesquisaFiltradasTools={ this.pesquisaFiltradasTools }
					toggleTagFiltro={ this.toggleTagFiltro }
				/>
				{modalAdicionarAberta &&
					<ModalVuttr fecharModal={() =>this.toggleModal()}>
						<AddTool 
							fecharModal={() =>this.toggleModal()}
							adicionarTool={ this.adicionarTool }
						/>
					</ModalVuttr>}
				{modalDeleteAberta &&
					<ModalVuttr fecharModal={() =>this.toggleModalDelete()}>
						<RemoveTool 
							fecharModal={() =>this.toggleModalDelete()}
							deleteTool={() =>this.deleteTool()}
							nomeTool={ nomeTool }
							isLoading={ isLoading }
							respostaFetchUsuario={ respostaFetchUsuario }
						/>
					</ModalVuttr>}
			</div>
		);
	}
}
export default App;

//Ideias: 
//Resetar input search box.-tentei
//Na pesquisa por tag fazer um highlight tna tags achadas.-tentei
//Quando adicionar mandar uma msg para o usuario.
//Mensagens de validação para erros de conexão, depois do loading do aperta do botão.
//Adicionar loading da lista.
//Falta fazer teste dos eventos addTag, removeTag e validação e loader do Addtool.js.
//Adicionar mensagem de lista vazia