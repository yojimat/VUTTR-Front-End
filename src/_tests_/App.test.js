import React from 'react';
import App from '../App';
import Renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';

const fetchCallMock = () => {
	const mockSuccessResponse = [{
				"id": 1,
      			"title": "Notion",
		        "link": "https://notion.so",
		        "description": "All in one tool to organize teams and ideas. Write, plan, collaborate, and get organized. ",
		        "tags": [
		        "organization",
		        "planning",
		        "collaboration",
		        "writing",
		        "calendar"
		        ]
			}],
    		mockJsonPromise = Promise.resolve(mockSuccessResponse),
    		mockFetchPromise = Promise.resolve({
      			json: () => mockJsonPromise,
    		});
    return mockFetchPromise;
}


describe(">>>App.js", () => {
	it("+App renderiza normalmente.", () => {
  		const appContainer = Renderer.create(<App />).toJSON();
  		expect(appContainer).toMatchSnapshot();
	});

	it("+Verificando ToggleModalAdd", () =>{
		const appToggle = mount(<App />);

		expect.assertions(3);
		expect(appToggle.state("modalAdicionarAberta")).toBe(false);
		appToggle.find("button").simulate('click');
		expect(appToggle.state("modalAdicionarAberta")).toBe(true);
		expect(appToggle.find(".addTool")).toHaveLength(1);
	});

	it("+Verificando ToggleModalRemove", () =>{
		const appToggleRemove = mount(<App />);

		expect.assertions(6);
		expect(appToggleRemove.state("toolsList")).toEqual([]);
		expect(appToggleRemove.state("modalDeleteAberta")).toBe(false);
		appToggleRemove.setState({ 
			toolsList: [{
				"id": 1,
      			"title": "Notion",
		        "link": "https://notion.so",
		        "description": "All in one tool to organize teams and ideas. Write, plan, collaborate, and get organized. ",
		        "tags": [
		        "organization",
		        "planning",
		        "collaboration",
		        "writing",
		        "calendar"
		        ]
			}]
		});
		expect(appToggleRemove.state("toolsList")).not.toEqual([]);
		expect(appToggleRemove.find("button[id=1]")).toHaveLength(1);
		appToggleRemove.find("button[id=1]").simulate('click');
		expect(appToggleRemove.state("modalDeleteAberta")).toBe(true);
		expect(appToggleRemove.find(".removeTool")).toHaveLength(1);
	});

	it("+Verificando se compenentDidMount chama getListaTools",() => {
		const spy = jest.spyOn(App.prototype, 'getListaTools'),
			appGetLista = mount(<App />);

		expect.assertions(2);
		expect(spy).toHaveBeenCalled();
		expect(spy).toHaveBeenCalledTimes(1);

		spy.mockClear();
	});

	it("+Verificando método getListaTools e state toolsList", done => {
  		const fetchCall = fetchCallMock(),
    		appGetListaTools = mount(<App />);

  		jest.spyOn(global, 'fetch').mockImplementation(() => fetchCall);

  		expect.assertions(4);
		expect(appGetListaTools.state().toolsList).toEqual([]);

		appGetListaTools.instance().getListaTools();

    	expect(global.fetch).toHaveBeenCalledWith("/tools",{"headers": {"Content-Type": "application/json"}, "method": "get"});
		expect(global.fetch).toHaveBeenCalledTimes(1);

		process.nextTick(() => {
			expect(appGetListaTools.state().toolsList).not.toEqual([]);

			global.fetch.mockClear();
			done();
		});	
	});

	it("+Verificando método getListaToolsBuscaGlobal", done =>{
		const fetchCallGlobal = fetchCallMock(),
    		appGetListaToolsGlobal = mount(<App />);

    	jest.spyOn(global, 'fetch').mockImplementation(() => fetchCallGlobal)

  		expect.assertions(4);
		expect(appGetListaToolsGlobal.state().toolsList).toEqual([]);

		appGetListaToolsGlobal.instance().getListaToolsBuscaGlobal("Notion");

    	expect(global.fetch).toHaveBeenCalledWith("/tools?q=Notion",{"headers": {"Content-Type": "application/json"}, "method": "get"});
    	//2 por causa do fetch do componentDidMount.
		expect(global.fetch).toHaveBeenCalledTimes(2);

		process.nextTick(() => {
			expect(appGetListaToolsGlobal.state().toolsList).not.toEqual([]);

			global.fetch.mockClear();
			done();
		});	
	});

	it("+Verificando método getListaToolsBuscaTag", done =>{
		const fetchCallTag = fetchCallMock(),
    		appGetListaToolsTag = mount(<App />);

    	jest.spyOn(global, 'fetch').mockImplementation(() => fetchCallTag)

  		expect.assertions(4);
		expect(appGetListaToolsTag.state().toolsList).toEqual([]);

		appGetListaToolsTag.instance().getListaToolsBuscaTag("organization");

    	expect(global.fetch).toHaveBeenNthCalledWith(2,"/tools?tags_like=organization",{"headers": {"Content-Type": "application/json"}, "method": "get"});
    	//2 por causa do fetch do componentDidMount.
		expect(global.fetch).toHaveBeenCalledTimes(2);

		process.nextTick(() => {
			expect(appGetListaToolsTag.state().toolsList).not.toEqual([]);

			global.fetch.mockClear();
			done();
		});	
	});

	it("+Verificando pesquisaFiltradasTools", () => {
		const appPesquisaFiltradasTools = mount(<App />),
			spyTag = jest.spyOn(App.prototype, 'getListaToolsBuscaTag'),
			spyGlobal = jest.spyOn(App.prototype, 'getListaToolsBuscaGlobal');

		expect(spyGlobal).toHaveBeenCalledTimes(0);

		appPesquisaFiltradasTools.instance().pesquisaFiltradasTools("jest");

		expect(spyGlobal).toHaveBeenCalled();
		expect(spyGlobal).toHaveBeenCalledTimes(1);

		expect(spyTag).toHaveBeenCalledTimes(0);

		appPesquisaFiltradasTools.setState({ tagFiltro: true });
		appPesquisaFiltradasTools.instance().pesquisaFiltradasTools("jest");

		expect(spyTag).toHaveBeenCalled();
		expect(spyTag).toHaveBeenCalledTimes(1);
	});

	it("+Verificando toggleTagFiltro", () => {
		const apptoggleTagFiltro = mount(<App />),
			spypesquisaFiltradasTools = jest.spyOn(apptoggleTagFiltro.instance(), 'pesquisaFiltradasTools');

		//Force Update para o jest entender que pesquisaFiltradasTools é uma função.
		apptoggleTagFiltro.instance().forceUpdate();

		expect(spypesquisaFiltradasTools).toHaveBeenCalledTimes(0);
		expect(apptoggleTagFiltro.state("tagFiltro")).toBe(false);

		apptoggleTagFiltro.instance().toggleTagFiltro("jest");

		expect(apptoggleTagFiltro.state("tagFiltro")).toBe(true);
		expect(spypesquisaFiltradasTools).toHaveBeenCalled();
		expect(spypesquisaFiltradasTools).toHaveBeenCalledTimes(1);
	});
});
