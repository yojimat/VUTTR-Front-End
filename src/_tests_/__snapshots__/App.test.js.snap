// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>>App.js +App renderiza normalmente. 1`] = `
<div>
  <div
    className="particles"
  >
    <canvas
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    />
  </div>
  <div
    className="App"
  >
    <header>
      <h1
        className="ink mt4 mb0 i fw1"
      >
        VUTTR
      </h1>
      <h2
        className="ink mt2 mb4 ttu tracked"
      >
        Very Useful Tools to Remember
      </h2>
    </header>
    <section>
      <form>
        <input
          aria-label="Pesquisa"
          className="mb2 pesquisaInput br1"
          onChange={[Function]}
          placeholder="Pesquise aqui…"
          type="search"
        />
         
        <label
          className="lh-copy container ml3"
        >
          Procurar com #tags?
          <input
            onChange={[Function]}
            type="checkbox"
          />
          <span
            className="checkmark br1"
          />
        </label>
      </form>
      <br />
      <button
        className="grow br2 ph3 pv2 mb2 dib bg-add azul"
        onClick={[Function]}
        type="button"
      >
        <span
          aria-label="plus"
          role="img"
        >
          ➕
        </span>
         Adicionar Tool
      </button>
      <br />
      <article
        className="bg-branco-escuro w-50 center mb4 o-90 ba b--black-20 br2 pr4 shadowBox"
      >
        <div
          className="pa2"
        >
          <span>
            Lista vazia.
          </span>
        </div>
      </article>
    </section>
  </div>
  <footer
    className="pb4 ph3 ph5-ns tc"
  >
    <a
      className="link near-black hover-silver dib h2 w2 mr3"
      href="https://github.com/yojimat"
      target="blank"
      title="GitHub"
    >
      <svg
        clipRule="evenodd"
        fill="currentColor"
        fillRule="evenodd"
        strokeLinejoin="round"
        strokeMiterlimit="1.414"
        viewBox="0 0 16 16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 0C3.58 0 0 3.582 0 8c0 3.535 2.292 6.533 5.47 7.59.4.075.547-.172.547-.385 0-.19-.007-.693-.01-1.36-2.226.483-2.695-1.073-2.695-1.073-.364-.924-.89-1.17-.89-1.17-.725-.496.056-.486.056-.486.803.056 1.225.824 1.225.824.714 1.223 1.873.87 2.33.665.072-.517.278-.87.507-1.07-1.777-.2-3.644-.888-3.644-3.953 0-.873.31-1.587.823-2.147-.083-.202-.358-1.015.077-2.117 0 0 .672-.215 2.2.82.638-.178 1.323-.266 2.003-.27.68.004 1.364.092 2.003.27 1.527-1.035 2.198-.82 2.198-.82.437 1.102.163 1.915.08 2.117.513.56.823 1.274.823 2.147 0 3.073-1.87 3.75-3.653 3.947.287.246.543.735.543 1.48 0 1.07-.01 1.933-.01 2.195 0 .215.144.463.55.385C13.71 14.53 16 11.534 16 8c0-4.418-3.582-8-8-8"
        />
      </svg>
    </a>
    <a
      className="link hover-silver near-black dib h2 w2 mr3"
      href="https://www.linkedin.com/in/vntc-fullstack/"
      target="blank"
      title="Linkedin"
    >
      <svg
        clipRule="evenodd"
        fill="currentColor"
        fillRule="evenodd"
        strokeLinejoin="round"
        strokeMiterlimit="1.414"
        viewBox="0 0 16 16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13.632 13.635h-2.37V9.922c0-.886-.018-2.025-1.234-2.025-1.235 0-1.424.964-1.424 1.96v3.778h-2.37V6H8.51V7.04h.03c.318-.6 1.092-1.233 2.247-1.233 2.4 0 2.845 1.58 2.845 3.637v4.188zM3.558 4.955c-.762 0-1.376-.617-1.376-1.377 0-.758.614-1.375 1.376-1.375.76 0 1.376.617 1.376 1.375 0 .76-.617 1.377-1.376 1.377zm1.188 8.68H2.37V6h2.376v7.635zM14.816 0H1.18C.528 0 0 .516 0 1.153v13.694C0 15.484.528 16 1.18 16h13.635c.652 0 1.185-.516 1.185-1.153V1.153C16 .516 15.467 0 14.815 0z"
          fillRule="nonzero"
        />
      </svg>
    </a>
    <br />
    <small>
      Projeto realizado para desafio do Bossabox.
    </small>
    <img
      alt="check"
      height="46"
      src="LogoBossabox.svg"
      width="36"
    />
  </footer>
</div>
`;
